packages/editor/src/components/wysiwyg/CodeLanguageSelect.tsx

```diff
@@ -1,20 +1,72 @@
-import React from "react"
+import { autoUpdate, flip, offset, useFloating } from "@floating-ui/react-dom"
+import React, { useEffect, useLayoutEffect } from "react"
 
 import { useCodeBlock } from "../../hooks/use-code-block"
 
+function getVirtualCodeBlock(top: number, left: number, right: number, bottom: number) {
+    const virtualCodeBlock = {
+        getBoundingClientRect() {
+            return {
+                x: left,
+                y: top,
+                top,
+                left,
+                bottom,
+                right,
+                width: right - left,
+                height: bottom - top,
+            }
+        },
+    }
+    return virtualCodeBlock
+}
+
 const CodeLanguageInput: React.FC = () => {
-    const { dom: reference, language, setLanguage } = useCodeBlock()
+    const { dom: codeBlock, language, setLanguage } = useCodeBlock()
+    const { top, left, right, bottom } = codeBlock
+        ? codeBlock.getBoundingClientRect()
+        : { top: -9999, left: -9999, right: -9999, bottom: -9999 }
+    const { x, y, reference, floating, strategy, update, refs } = useFloating({
+        placement: "top-end",
+        middleware: [flip(), offset(1)],
+    })
 
-    if (reference) {
-        const { top, left, right, bottom } = reference.getBoundingClientRect()
-        console.log("[CodeLanguageSelect] referance:", { top, left, right, bottom })
-        console.log("[CodeLanguageSelect] language:", language)
-    }
+    useEffect(() => {
+        if (!refs.reference.current || !refs.floating.current) {
+            return
+        }
+        // Only call this when the floating element is rendered
+        return autoUpdate(refs.reference.current, refs.floating.current, update)
+    }, [refs.reference, refs.floating, update])
+
+    useLayoutEffect(() => {
+        // Call reference with the virtual element inside an effect
+        reference(getVirtualCodeBlock(top, left, right, bottom))
```

Comment:
It seems that we can just pass `codeBlock` (instead of a virtual element) here. I don't know why I trend to use a virtual element in the first place. `codeBlock` is a HTMLElement or `null`, and `reference` accepts `null`, which make thing easier. 

We can use `if (!codeBlock) {return null} else {return <input...>}` here to get ride of -9999.